### ISSUE TITLE
Make Driver Cache thread-safe

### ISSUE THREAD

**Issue by Kirill888** [opendatacube/datacube-core#624]
There are 3 driver cache singletons: Read, Write and Index, all assume that first access is done without concurrency as they have code like this:

https://github.com/opendatacube/datacube-core/blob/c4595b91387011d4deec3c2d63f88e9e9aa01573/datacube/drivers/readers.py#L40-L46

Note that line `45` might end up executing concurrently in two different threads.

**Comment by petewa** [comment]
We encountered a problem on a race condition on simultaneous reader/writer/index cache creation when the cache is first initialized/created in multiple threads nearly a year ago. I shared this fix with GA when we discovered the issue but just in case it is still relevant.

This is how we are handling that issue in the csiro/execution-engine branch.
The fix is also in a branch: https://github.com/opendatacube/datacube-core/tree/csiro/develop-fix-drivers @ cfc2fa4ef5ccc2772351666c251690dfc3d6e29c

It solves one of the concurrency issues, there are a few others in core, some in the driver cache.

Other parts of the cache could use locking as well, but fixes there were not applied to maintain as much compatibility with core as possible but we are running into more concurrency issues now with more scaling.

Our fix above is a stop gap, and a proper fix to promote thread-safety should be done.

**Comment by Kirill888** [comment]
@petewa I created this issue after stalking your fork, do you mind expanding more on other concurrency issues you have experienced. I do not recall seeing issue/PR about driver caches, but like you said this is from march last year...

**Comment by Kirill888** [comment]
I see here it is:

https://github.com/opendatacube/datacube-core/issues/323#issuecomment-398939108

I certainly missed it back then, in the future would be nice to have these kind of problemes raised as proper issues so they won't go missing for a year or so.

**Comment by petewa** [comment]
@Kirill888 I don't think I created an issue if memory serves. I mentioned in a meeting with yourself and Damien and in the SC. My memory is very hazy here.

The issue from a year ago was during the first dc.load(), many threads would attempt to simultaneously try to create the cache and it would deadlock? I think, don't recall. It presented itself in pytests as well.

The concurrency issue I'm trying to solve right now is, threadpool/processpool concurrency on creation, which is influenced by the cache. The way I did it was multi-tier, thread on time in core and space in the s3 driver. It was done like this to maintain compatibility with core, I somewhat regret this decision in hindsight. I should have done it properly even though it would cause more pain for GA during the s3 driver PR.

I'm thinking of moving the threadpool/process pool up further into core, possibly using/extending the executor class with pathos, but to do this properly there needs to be some sort of IO planning to prevent deadlocks.

For example:
for a dc.load():
* dc.metadata_for_load() - get all metadata for a dc.load
* dc.load_data()
  * dc.create_storage - pre-create one contiguous array to hold all data.
  * Plan all IO and produce a list of work items i.e. ([t, x, y, band, etc IO read] -> array slice)
  * Send this to the core threadpool/processpool
  * All work items happen in parallel and write/streams back into the pre-created array in parallel.
  * pre-created array may be a shared memory array that may be in memory or disk and is accessible across threads & processed.

At least that's the direction I'm heading towards. I try to think of every function in an array operator sort of way so everything comes down to an array operator of sorts.

Whether I get time to work on it is a different story.

The above is similar to what happens in our Job Result Object (JRO) in our execution-engine branch for future computed results. It creates an array mapping that maps array slices to a list actual IO calls that can stream in parallel back to the result array.

I'll stop there, too much information to digest already.

**Comment by woodcockr** [comment]
@petewa Would be worth expanding your thoughts on the dc.save() threading as well since we have that on the futures path as well.

**Comment by petewa** [comment]
@woodcockr, ahh yes dc.save() https://github.com/opendatacube/datacube-core/tree/datacube-save
#467 
The code above works, with a few caveats and can be improved upon further.

My thoughts can be hard to follow :) I have trouble myself.

My thoughts is to make dc.save() happen in parallel with a simple array like call.
Either dc.save a whole xarray dataset/dataarray or a subset of an array (array slice).

For reads, s3aio (what the s3-driver uses) maps an array slice to a list of read IO calls to s3.
For writes, we can reuse that mapping to write back to s3 in parallel, because the small 5-10mb s3aio chunks/blocks are stored in different s3 objects, writes *should be* amazingly parallel and fast.
s3aio was designed for fast reads & writes (not implemented yet, function stub is there) given s3's limitation of a single change to a s3 object, re-writes the entire object. This is why we went with a chunk per s3 object, trading off some IO $ cost for more potential write performance.

s3aio *could* also be standalone library that maps a big array, sharded to s3 with parallel reads and writes via standard array slice operators.

Massively parallel writes back to s3, is what may set us a part, if done right.

If done right, it can do a parallel save across nodes as well and could also be used for easy ingest.

For example:
my big virtual array x[100000000000, 100000000000, 1000000000000, 1000000000000]
x is mapped across s3.
I want to save/ingest that into another form
I can do a dc.save() on x and it could operate on sub-slices per node
The final array shares the same mapping (very convenient) and can be indexed as usual with core like it happens now with the s3-driver.

Within a single node, it'll use the same threading model in the post further up.

Would anyone like to see this happen?

**Comment by omad** [comment]
Hi @petewa, thanks for the detailed posts! Could you please move your comment into #625 (_Discussion: Concurrent IO_) so that we can make this discussion more visible and discoverable.
@Kirill888 is going to post our thoughts and planned direction on concurrent IO there.

Briefly, we're moving towards making concurrency purely a driver issue, and remove threading/process pools from the core code.

**Comment by petewa** [comment]
@omad moving the concurrency towards the driver is also possible but currently the drivers do not have enough information to make it work well and fuse_measurements will need to be decoupled out as a extra post_process step after the IO. I'll duplicate the post in #625.
